cpp/DEVELOPER_GUIDE.md

```diff
@@ -5,9 +5,22 @@ This document summarizes rules and best practices for contributions to the cuML
 Please start by reading [CONTRIBUTING.md](../CONTRIBUTING.md).
 
 ## Thread safety
-cuML is thread safe so its functions can be called from multiple host threads if they use different handles.
 
-The implementation of cuML is single threaded.
+With the exception of the cumlHandle, cuML algorithms should maintain thread-safety and are, in general, 
+assumed to be single threaded. This means they should be able to be called from multiple host threads so 
+long as different instances of `cumlHandle` are used.
+
+Exceptions are made for algorithms that can take advantage of multiple CUDA streams in order to oversubscribe 
+or increase occupancy on a single GPU. In these cases, the use of multiple host threads within cuML algorithms
+should be used only to maintain concurrency of the underlying CUDA streams. Multiple host threads should be 
+used sparingly, be bounded, and should steer clear of performing CPU-intensive computations.
+
+A good example of an acceptable use of host threads can be found in [this blog article](https://devblogs.nvidia.com/gpu-pro-tip-cuda-7-streams-simplify-concurrency/)
+
+Though cuML's build enables OpenMP by default, cuML algorithms should still function properly even when OpenMP has been
+disabled. For this reason, it would be better to utilize OpenMP rather than raw pthreads for spawning host threads. 
+
+The use of threads in thirdparty libraries are okay.
```

Comment:
```suggestion
The use of threads in third-party libraries is okay.
```

Reply:
If this sentence should go in I think it needs to be more specific. E.g. one reason why it is desirable to have OpenMP only as optional dependency is to avoid compatibility issues with a OpenMP runtime a cuML might want to use. E.g. if the application calling into cuML spawns multiple OpenMP threads and let each OpenMP thread work on independent cuML jobs. For that reason third party libraries should avoid to in a dependency for a specific OpenMP runtime.